Userland/Libraries/LibWeb/Painting/CheckBoxPaintable.cpp

```diff
@@ -33,4 +90,72 @@ Layout::CheckBox& CheckBoxPaintable::layout_box()
     return static_cast<Layout::CheckBox&>(layout_node());
 }
 
+void CheckBoxPaintable::paint(PaintContext& context, PaintPhase phase) const
+{
+    if (!is_visible())
+        return;
+
+    PaintableBox::paint(context, phase);
+
+    if (phase == PaintPhase::Foreground) {
+        auto const& checkbox = static_cast<HTML::HTMLInputElement const&>(layout_box().dom_node());
+        bool enabled = layout_box().dom_node().enabled();
+        Gfx::AntiAliasingPainter painter { context.painter() };
+        auto checkbox_rect = context.enclosing_device_rect(absolute_rect()).to_type<int>();
+        auto checkbox_radius = checkbox_rect.width() / 5;
+
+        auto& palette = context.palette();
+
+        auto shade = [&](Color color, float amount = 0.3f) {
+            return color.mixed_with(palette.is_dark() ? Color::Black : Color::White, amount);
+        };
+
+        auto modify_color = [&](Color color) {
+            if (being_pressed() && enabled)
+                return shade(color);
+            return color;
+        };
+
+        auto base_text_color = palette.color(ColorRole::BaseText);
+        auto accent = computed_values().accent_color().value_or(palette.color(ColorRole::Accent));
+        auto base = shade(base_text_color.inverted(), 0.8f);
+        auto dark_gray = shade(base_text_color);
+        auto gray = shade(dark_gray, 0.4f);
+        auto mid_gray = shade(gray);
+        auto light_gray = shade(mid_gray);
+
+        auto increase_contrast = [&](Color color, Color background) {
+            auto constexpr min_contrast = 2;
+            if (color.contrast_ratio(background) < min_contrast) {
+                color = color.inverted();
+                if (color.contrast_ratio(background) > min_contrast)
+                    return color;
+            }
+            return color;
+        };
+
+        float smoothness = 1.0f / (max(checkbox_rect.width(), checkbox_rect.height()) / 2);
+        if (checkbox.checked()) {
```

Comment:
One small thing. The indeterminate state should take precedence over checked: 
> If the element's [indeterminate](https://html.spec.whatwg.org/multipage/input.html#dom-input-indeterminate) IDL attribute is set to true, then the control's selection should be obscured as if the control was in a third, indeterminate, state.

[spec](https://html.spec.whatwg.org/multipage/input.html#checkbox-state-(type=checkbox):dom-input-indeterminate)

Reply:
Ah, thank you!